### PR TITLE
Add Trino fault tolerance architecture post

### PR DESCRIPTION
**Pull request by bitsondatadev**

### CODE REVIEW

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+stable and predictable landing time. Moreover, it opens the opportunity for
+running on much cheaper spot instances as well as auto scaling our Trino
+clusters to further optimize costs.
+
+## Resource Management
+
+Buffering intermediate results makes it possible to execute tasks iteratively,
+and we are no longer constrained by the size of the cluster to configure the
+number of partitions. This greatly reduces the memory pressure of large queries,
+as now we can process a subset of data at a time. Executing memory intensive
```
Comment (linzebing) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:126]:
```suggestion
as now we can process a subset of data at a time. Executing memory-intensive
```

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+#### Exchange Spooling
+
+The intermediate storage for spooling data should be able to buffer, store and
+return it efficiently.
+
+![](/assets/blog/architecting-trino-next-gen-etl-engine/exchange-spooling.png)
+
+Currently, we have implemented a filesystem-based approach in the 
+`trino-exchange` plugin, supporting local filesystem, AWS S3, Azure Blob
```
Comment (linzebing) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:103]: It has been renamed already. Also local filesystem is misleading --- it's not applicable to multiple workers.
```suggestion
`trino-exchange-filesystem` plugin, supporting AWS S3, Azure Blob
```

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+Storage, Google Cloud Storage, as well as S3-compatible systems. The
+implementation takes full advantage of S3 asynchronous APIs to overlap
```
Comment (linzebing) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:105]:
```suggestion
implementation takes full advantage of asynchronous APIs to overlap
```

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+computation and I/O, and performs parallel reads and writes from and to exchange
+storage in parts to maximize throughput. We also opportunistically coalesce and
+split partition files depending on the size. 
+
+While the existing exchange spooling layer currently uses S3, we designed the
+`trino-exchange` plugin to allow for the community to build a distributed
+buffering service to take advantage of write ahead buffers that aggregate data
+belonging to the same partition. We envision this could be backed by a hierarchy
+of storage tiers to further enhance exchange performance.
```
Comment (linzebing) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:114]:
```suggestion
The exchange APIs defined in the Trino SPI are designed to be extensible, and allow for the community to build a distributed buffering service to further enhance exchange performance.
```

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+Currently, we have implemented a filesystem-based approach in the 
+`trino-exchange` plugin, supporting local filesystem, AWS S3, Azure Blob
+Storage, Google Cloud Storage, as well as S3-compatible systems. The
+implementation takes full advantage of S3 asynchronous APIs to overlap
+computation and I/O, and performs parallel reads and writes from and to exchange
+storage in parts to maximize throughput. We also opportunistically coalesce and
+split partition files depending on the size. 
+
+While the existing exchange spooling layer currently uses S3, we designed the
+`trino-exchange` plugin to allow for the community to build a distributed
```
Comment (bitsondatadev) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:111]:
```suggestion
`trino-exchange-filesystem` plugin to allow for the community to build a distributed
```

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+# A new execution paradigm
+
+![](/assets/blog/architecting-trino-next-gen-etl-engine/current-architecture.png)
+
+To address this, we introduce a new execution paradigm to Trino by adding
+exchange spooling that stores intermediate data in an efficient buffering layer
+at stage boundaries.This opens a lot of new opportunities around failure
+recovery, resource management, and query optimization.
+
+## Fault tolerance
```
Comment (bitsondatadev) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:59]:
```suggestion
## Fault-tolerance
```

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+There are now two levels of fault tolerance into Trino:
```
Comment (bitsondatadev) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:61]:
```suggestion
There are now two levels of fault-tolerance into Trino:
```

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+is integrating the existing
+[dynamic filtering and dynamic partition pruning](https://trino.io/episodes/11.html)
+features with the iterative task execution paradigm.
+
+This is a space full of innovations, and we are just getting started.
+
+# Conclusion
+
+Project Tardigrade brought a new execution paradigm into Trino, which provides
+first-class support for batch processing. Featuring fault tolerance and flexible
```
Comment (bitsondatadev) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:162]:
```suggestion
first-class support for batch processing. Featuring fault-tolerance and flexible
```

**Review by linzebing — APPROVED** [review]

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+isolation, speculative execution, adaptive query planning, and skew handling to
+scale Trino worker execution.
+
+# A new execution paradigm
+
+![](/assets/blog/architecting-trino-next-gen-etl-engine/current-architecture.png)
+
+To address this, we introduce a new execution paradigm to Trino by adding
+exchange spooling that stores intermediate data in an efficient buffering layer
+at stage boundaries.This opens up a lot of new opportunities around failure
```
Comment (martint) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:56]: Space after period

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+Currently, we have implemented a filesystem-based approach in the 
+`trino-exchange-filesystem` plugin, supporting AWS S3, Azure Blob
+Storage, Google Cloud Storage, as well as S3-compatible systems. The
+implementation takes full advantage of asynchronous APIs to overlap
+computation and I/O, and performs parallel reads and writes from and to exchange
+storage in parts to maximize throughput. We also opportunistically coalesce and
+split partition files depending on the size. 
+
+The exchange APIs defined in the Trino SPI are designed to be extensible, and allow for the community to build a distributed buffering service to further enhance exchange performance.
```
Comment (martint) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:110]: > and allow for the community to build 

It's not "us vs the community". I would rephrase to explain that this allows for customized implementations of the exchange spooling APIs, and that future directions might involve building a dedicated buffering service using tiered storage (memory, ssd, object store)

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+With fault tolerance built into Trino, long-running batch queries now can enjoy
```
Comment (martint) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:112]: This is not quite true. Stable and predictable completion times depend on many other factors than granular retries (concurrent workloads, resource management, performance of dependent systems, etc).

Also, "landing time" is a very vague term (specifically used within FB), but I don't know if it's in common usage in the industry. Is it defined elsewhere?

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+featuring in-memory streaming shuffle. Tasks are interconnected and data is
+streamed through the task dependency graph with no intermediate checkpointing.
+This all-or-nothing architecture makes it really hard to tolerate faults. If
+any of the tasks dies/fails, there is no way to resume execution from some
+intermediate state, the only option is to restart the whole query. This is
+extremely painful for long-running queries as they are more likely to fail, and
+impact of failures is much higher as it often results in significant waste of
+time and resources.
+
+With the exception of grouped execution,which applies to limited scenarios like
```
Comment (martint) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:37]: Space after comma

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+at stage boundaries. This opens up a lot of new opportunities around failure
+recovery, resource management, and query optimization.
+
+## Fault-tolerance
+
+There are now two levels of fault-tolerance into Trino:
+
+### Query-level Retry
+
+By maintaining a deduplication buffer with the ability to spill to external
```
Comment (martint) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:65]: "deduplication buffer" needs some clarification. It's not obvious what that means for someone unfamiliar with the implementation

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+# Conclusion
+
+Project Tardigrade brought a new execution paradigm into Trino, which provides
+first-class support for batch processing. Featuring fault-tolerance and flexible
+scheduling, it greatly improves Trino’s support for long-running and
+memory-intensive queries. To learn how we evaluated performance, you can find
+[our benchmark results on TPC-H data](https://www.starburst.io/blog/a-better-solution-for-managing-and-maintaining-data-pipelines-now-in-public-preview/).
+
+We are excited for the Trino community to try it out. We provide a one-click
+option for enabling it out using a a [fully managed Trino (Starburst Galaxy)](https://www.starburst.io/platform/starburst-galaxy/),
```
Comment (mosabua) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:168]: sorry ... what does this sentence want to say?

File: _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+or you can [try out the tutorial using open source Trino](https://github.com/bitsondatadev/trino-getting-started/tree/main/kubernetes/tardigrade-eks).
+Please reach out to us on [Trino Slack](https://trino.io/slack.html) in the
```
Comment (mosabua) [inline _posts/2022-08-10-architecting-trino-next-gen-etl-engine.md:170]: Drop "Please"

**Review by mosabua — CHANGES_REQUESTED** [review]
Overall there is a lot of good content in here but the post misses an arc for a story line. Ping me to chat and discuss. Might be a bigger refactor. We need to figure out what we want to communicate .. currently that is not clear to me.

**Comment by bitsondatadev** [comment]
Oops...Mah bad

**Comment by mosabua** [comment]
I am closing this .. feel free to reopen if you end up working on it at some stage.
